server/src/test/java/org/elasticsearch/action/search/SearchAsyncActionTests.java

```diff
@@ -161,28 +158,18 @@ public void testLimitConcurrentShardRequests() throws InterruptedException {
         request.allowPartialSearchResults(true);
         int numConcurrent = randomIntBetween(1, 5);
         request.setMaxConcurrentShardRequests(numConcurrent);
-        CountDownLatch latch = new CountDownLatch(1);
-        AtomicReference<TestSearchResponse> response = new AtomicReference<>();
-        ActionListener<SearchResponse> responseListener = new ActionListener<SearchResponse>() {
-            @Override
-            public void onResponse(SearchResponse searchResponse) {
-                response.set((TestSearchResponse) searchResponse);
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                logger.warn("test failed", e);
-                fail(e.getMessage());
-            }
-        };
+        int numShards = 10;
+        CountDownLatch latch = new CountDownLatch(numShards);
+        AtomicBoolean searchPhaseDidRun = new AtomicBoolean(false);
+        ActionListener<SearchResponse> responseListener = ActionListener.wrap(response -> {}, e -> fail(e.getMessage()));
```

Comment:
`fail(e.getMessage())` loses the stack trace from the thrown exception, which might be useful. How about `throw new AssertionError("unexpected", e);` instead?